### PR TITLE
feat(healthcheck, openebs operator): introduce openebs operator - drop1

### PR DESCRIPTION
**Pull request by AmitKumarDas**
This commit introduces _idiomatic maya_ as a series of commit to implement openebs operator

- Idiomatic maya will be the core layer of OpenEBS control plane
- Features including CASTemplate, Operators, etc are being built or re-built on top of idiomatic maya
- More than features, this becomes the language of team communication (internal as well as external)
- Bugs are more evident during the development phase itself
- This reduces the time taken to review
- This reduces the dependency on individual team members
- Team starts to talk w.r.t patterns (it should really become boring & hence quick to implement)
- This reduces the effort required to maintain code
- This improves the time taken to mark a feature as complete

This commit includes:
- Code against _runtask_, _daemonset_, _container_, _toleration_ using idiomatic maya principles
- Code that uses kubernetes client-go using idiomatic maya principles

### CODE REVIEW

File: pkg/apis/openebs.io/v1alpha1/register.go [hunk 1]
```diff
@@ -1,11 +1,10 @@
 package v1alpha1
```
Comment (ashishranjan738) [inline pkg/apis/openebs.io/v1alpha1/register.go]: add copyright info

File: pkg/kubernetes/client/v1alpha1/client.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	K8sMasterIPEnvironmentKey env.ENVKey = "OPENEBS_IO_K8S_MASTER"
+	// KubeConfigEnvironmentKey is the environment variable key used to
+	// determine the kubernetes config
+	KubeConfigEnvironmentKey env.ENVKey = "OPENEBS_IO_KUBE_CONFIG"
+)
+
+// getInClusterConfigFunc is a typed function to get incluster config
+//
+// NOTE:
+//  functional makes it simple to mock
```
Comment (sonasingh46) [inline pkg/kubernetes/client/v1alpha1/client.go]: I think -- this comment can be improved.
Got confused what `functional` can mean here.

File: pkg/kubernetes/client/v1alpha1/client.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+type getKubeConfigPathFunc func(env.ENVKey) string
+
+// getKubernetesClientsetFunc is a typed function to get kubernetes
+// clientset
+//
+// NOTE:
+//  functional makes it simple to mock
+type getKubernetesClientsetFunc func(*rest.Config) (*kubernetes.Clientset, error)
+
+// GetConfigFunc is a typed function to get kubernetes config
```
Comment (sonasingh46) [inline pkg/kubernetes/client/v1alpha1/client.go]: Redundant comment -- Line no 43
How these two are different?

File: pkg/kubernetes/client/v1alpha1/client.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+// config returns the kubernetes config instance based on available criteria
+func (c *Client) Config() (config *rest.Config, err error) {
+	if c.IsInCluster {
+		return c.getInClusterConfig()
+	}
+	if strings.TrimSpace(c.getKubeMasterIP(K8sMasterIPEnvironmentKey)) != "" ||
+		strings.TrimSpace(c.getKubeConfigPath(KubeConfigEnvironmentKey)) != "" {
+		return c.getConfigFromENV()
+	}
+	return c.getInClusterConfig()
```
Comment (sonasingh46) [inline pkg/kubernetes/client/v1alpha1/client.go]: Either this or `if` block can be removed.

File: pkg/kubernetes/client/v1alpha1/client.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+		c.IsInCluster = true
+	}
+}
+
+// config returns the kubernetes config instance based on available criteria
+func (c *Client) Config() (config *rest.Config, err error) {
+	if c.IsInCluster {
+		return c.getInClusterConfig()
+	}
+	if strings.TrimSpace(c.getKubeMasterIP(K8sMasterIPEnvironmentKey)) != "" ||
```
Comment (sonasingh46) [inline pkg/kubernetes/client/v1alpha1/client.go]: I think we should replace `||` by `&&`

File: pkg/kubernetes/client/v1alpha1/client.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+	}
+	return c.getInClusterConfig()
+}
+
+func (c *Client) getConfigFromENV() (config *rest.Config, err error) {
+	k8sMaster := c.getKubeMasterIP(K8sMasterIPEnvironmentKey)
+	kubeConfig := c.getKubeConfigPath(KubeConfigEnvironmentKey)
+	if strings.TrimSpace(k8sMaster) == "" &&
+		strings.TrimSpace(kubeConfig) == "" {
+		return nil, errors.Errorf("failed to get kubernetes config: missing environment variables: atleast one should be set: %s or %s", K8sMasterIPEnvironmentKey, KubeConfigEnvironmentKey)
```
Comment (sonasingh46) [inline pkg/kubernetes/client/v1alpha1/client.go]: Error message looks ambiguous to me.  -- `atleast one should be set: %s or %s", K8sMasterIPEnvironmentKey, KubeConfigEnvironmentKey`
But if condition says both should be set.

File: pkg/kubernetes/client/v1alpha1/client.go [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+	}
+	return c.buildConfigFromFlags(k8sMaster, kubeConfig)
```
Comment (sonasingh46) [inline pkg/kubernetes/client/v1alpha1/client.go]: We have handled the empty values for `k8sMaster` and `kubeConfig` (line no 151) but it is already checked before calling this method. That means there could be multiple callers --may be our intention is that we want this function(`getConfigFromENV` function) to be used by different callers and error handling should be taken care here in the `getConfigFromENV` function. Sounds good.
Now -- 1. If we want this function to called by different callers then on this principle( stated above) `c.buildConfigFromFlags` should be also checked for its empty value or for both cases it should bot be checked.

File: pkg/kubernetes/client/v1alpha1/client.go [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+//
+// NOTE:
+//  functional makes it simple to mock
+type GetConfigFunc func(*Client) (*rest.Config, error)
+
+// GetConfig returns kubernetes config instance
+//
+// NOTE:
+//  This is an implementation of GetConfigFunc
+func GetConfig(c *Client) (*rest.Config, error) {
```
Comment (sonasingh46) [inline pkg/kubernetes/client/v1alpha1/client.go]: I am not sure where we are using this function. Looks like if we strip of this code there would be no impact.
What is the reason we want to have this piece of code.

File: pkg/apis/openebs.io/runtask/v1beta1/runtask.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+// resource
+type Action string
+
+const (
+	Get      Action = "get"
+	Template Action = "template"
+	List     Action = "list"
+	Create   Action = "create"
+	Update   Action = "update"
+	Patch    Action = "patch"
```
Comment (ashishranjan738) [inline pkg/apis/openebs.io/runtask/v1beta1/runtask.go]: is put action getting removed ?

File: pkg/kubernetes/container/v1alpha1/container.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+	return b.con.asContainer(), nil
+}
+
+// AddCheck adds the predicate as a condition to be validated against the
+// container instance
+func (b *builder) AddCheck(p Predicate) *builder {
+	b.checks = append(b.checks, p)
+	return b
+}
+
```
Comment (sonasingh46) [inline pkg/kubernetes/container/v1alpha1/container.go]: Return of the function is not used

File: pkg/kubernetes/container/v1alpha1/container.go [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+// AddChecks adds the provided predicates as conditions to be validated against
+// the container instance
+func (b *builder) AddChecks(p []Predicate) *builder {
+	for _, check := range p {
+		b.AddCheck(check)
+	}
+	return b
+}
+
```
Comment (sonasingh46) [inline pkg/kubernetes/container/v1alpha1/container.go]: Why is the additional `AddCheck` function required ? IMO -- we can have following
```
func (b *builder) AddChecks(p []Predicate) *builder {
	for _, check := range p {
		b.checks = append(b.checks, p)
	}
	return b
}
```
Similarly, at other places.

File: pkg/kubernetes/container/v1alpha1/container.go [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+// builder provides utilities required to build a kubernetes container type
+type builder struct {
+	con    *container  // container instance
+	checks []Predicate // validations to be done while building the container instance
+	errors []error     // errors found while building the container instance
+}
+
+// Builder returns a new instance of builder
```
Comment (sonasingh46) [inline pkg/kubernetes/container/v1alpha1/container.go]: `// Builder returns a new instance of builder`  --> `// Builder returns a new empty instance of builder`

File: pkg/kubernetes/client/v1alpha1/client.go (hunk 2 above)
Comment (AmitKumarDas) [reply pkg/kubernetes/client/v1alpha1/client.go]: done

File: pkg/kubernetes/client/v1alpha1/client.go (hunk 5 above)
Comment (AmitKumarDas) [reply pkg/kubernetes/client/v1alpha1/client.go]: as discussed || is correctly used here

File: pkg/kubernetes/client/v1alpha1/client.go (hunk 3 above)
Comment (AmitKumarDas) [reply pkg/kubernetes/client/v1alpha1/client.go]: will address this

File: pkg/kubernetes/client/v1alpha1/client.go [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+)
+
+// getInClusterConfigFunc is a typed function to get incluster config
+//
+// NOTE:
+//  functional makes it simple to mock
+type getInClusterConfigFunc func() (*rest.Config, error)
+
+// buildConfigFromFlagsFunc is a typed function to get desired
+// kubernetes config
```
[the remark this answers is not in the document]
Comment (AmitKumarDas) [reply pkg/kubernetes/client/v1alpha1/client.go]: will address this

File: pkg/kubernetes/client/v1alpha1/client.go [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+//
+// NOTE:
+//  functional makes it simple to mock
```
[the remark this answers is not in the document]
Comment (AmitKumarDas) [reply pkg/kubernetes/client/v1alpha1/client.go]: will address this

File: pkg/kubernetes/client/v1alpha1/client.go (hunk 7 above)
Comment (AmitKumarDas) [reply pkg/kubernetes/client/v1alpha1/client.go]: validations are present in each method as they might be called by different callers.

File: pkg/kubernetes/client/v1alpha1/client.go (hunk 8 above)
Comment (AmitKumarDas) [reply pkg/kubernetes/client/v1alpha1/client.go]: This is exposed to be used by other packages.

File: pkg/apis/openebs.io/runtask/v1beta1/runtask.go (hunk 9 above)
Comment (AmitKumarDas) [reply pkg/apis/openebs.io/runtask/v1beta1/runtask.go]: We are using create to create. Put will be used when update is supported.
Inn other words, put will be used for update cases.

File: pkg/kubernetes/container/v1alpha1/container.go (hunk 12 above)
Comment (AmitKumarDas) [reply pkg/kubernetes/container/v1alpha1/container.go]: I am focusing more on what it provides versus what it contains. Need not be followed as a rule though.

File: pkg/kubernetes/container/v1alpha1/container.go (hunk 10 above)
Comment (AmitKumarDas) [reply pkg/kubernetes/container/v1alpha1/container.go]: I did not understand the comment

File: pkg/runtask/v1beta1/runtask.go [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+	return b.runtask.task, nil
+}
+
+// AddCheck adds the predicate as a condition to be validated against the
+// runtask instance
+func (b *builder) AddCheck(p Predicate) *builder {
+	b.checks = append(b.checks, p)
+	return b
+}
+
```
[the remark this answers is not in the document]
Comment (AmitKumarDas) [reply pkg/runtask/v1beta1/runtask.go]: Could not understand the comment

File: pkg/runtask/v1beta1/runtask.go [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+// AddChecks adds the provided predicates as conditions to be validated against
+// the runtask instance
+func (b *builder) AddChecks(p []Predicate) *builder {
+	for _, check := range p {
+		b.AddCheck(check)
+	}
+	return b
+}
+
```
[the remark this answers is not in the document]
Comment (AmitKumarDas) [reply pkg/runtask/v1beta1/runtask.go]: Have explained this in earlier comment

File: pkg/kubernetes/client/v1alpha1/client.go (hunk 6 above)
Comment (AmitKumarDas) [reply pkg/kubernetes/client/v1alpha1/client.go]: This is correct. At least one of the arguments is required.

File: pkg/kubernetes/daemonset/v1alpha1/daemonset.go [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+// AddNodeSelector adds the provided pair as a nodeselector to the daemonset
+func AddNodeSelector(k, v string) OptionFunc {
+	return func(d *daemonset) {
+		if d.daemon.Spec.Template.Spec.NodeSelector == nil {
+			d.daemon.Spec.Template.Spec.NodeSelector = map[string]string{}
+		}
+		d.daemon.Spec.Template.Spec.NodeSelector[k] = v
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (sonasingh46) [reply pkg/kubernetes/daemonset/v1alpha1/daemonset.go]: Looks good as discussed

File: pkg/kubernetes/container/v1alpha1/container.go (hunk 10 above)
Comment (sonasingh46) [reply pkg/kubernetes/container/v1alpha1/container.go]: Considering this as a part of next PR.

File: pkg/runtask/v1beta1/runtask.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (sonasingh46) [reply pkg/runtask/v1beta1/runtask.go]: Considering this as part of next PR

File: pkg/client/generated/openebs.io/runtask/v1beta1/informer/externalversions/openebs.io/interface.go [hunk 18]
```diff
@@ -0,0 +1,46 @@
+/*
```
Comment (prateekpandey14) [inline pkg/client/generated/openebs.io/runtask/v1beta1/informer/externalversions/openebs.io/interface.go]: I think we should handle generated code simply .. ? too much versioning IMO in auto gen code ?

File: pkg/client/generated/openebs.io/runtask/v1beta1/informer/externalversions/openebs.io/interface.go [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+*/
+
+// Code generated by informer-gen. DO NOT EDIT.
+
+package openebs
+
+import (
+	internalinterfaces "github.com/openebs/maya/pkg/client/generated/openebs.io/runtask/v1beta1/informer/externalversions/internalinterfaces"
+	v1alpha1 "github.com/openebs/maya/pkg/client/generated/openebs.io/runtask/v1beta1/informer/externalversions/openebs.io/v1alpha1"
+)
```
Comment (prateekpandey14) [inline pkg/client/generated/openebs.io/runtask/v1beta1/informer/externalversions/openebs.io/interface.go]: Having complex auto-gen `informer and listers` path will be difficult to manage ( too long and redundant )?

Comment (AmitKumarDas) [reply pkg/client/generated/openebs.io/runtask/v1beta1/informer/externalversions/openebs.io/interface.go]: This is a good point. Not sure how to handle this.
I am not sure why it refers to externalversions?
Can we avoid externalversions?
@prateekpandey14

File: pkg/client/generated/openebs.io/runtask/v1beta1/informer/externalversions/openebs.io/interface.go (hunk 18 above)
Comment (AmitKumarDas) [reply pkg/client/generated/openebs.io/runtask/v1beta1/informer/externalversions/openebs.io/interface.go]: The pkg structure has been thought off extensively.
We have taken openebs code into consideration versus. how it look once code gen happens.
Is there a way to suppress externalversions? @prateekpandey14

File: pkg/runtask/v1beta1/kubernetes_client.go [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+func getRuntask(k *kubeclient, cs *internalclientset.Clientset, name string) (*apis.RunTask, error) {
+	if k == nil {
+		return nil, errors.New("failed to get runtask: nil kubeclient was provided")
+	}
+	if cs == nil {
+		return nil, errors.New("failed to get runtask: nil clientset was provided")
+	}
+	return cs.OpenebsV1beta1().RunTasks(k.namespace).Get(name, v1.GetOptions{})
+}
+
```
Comment (sonasingh46) [inline pkg/runtask/v1beta1/kubernetes_client.go]: For the above -- what is resource is clustered scope ? In this case, we do not require `k` .
Why both `kubeclient` and `internalclient` are used here ?

**Pull request merged**